### PR TITLE
Allow user defined Call-ID to be sent on outgoing INVITE

### PR DESCRIPTION
**Pull request by austinkottke**
Modify the call setting structure to allow a custom Call-ID which does not depend on pjsip's dynamic pj_create_unique_string -- make_call will use the predefined Call-ID if specified. 

Instead it allows the app developer to have control. 

Example usage:

```
call_opt.custom_call_id = pj_str("0192983-28839-23287892");

pjsua_call_make_call(current_acc, &tmp, &call_opt, NULL, &msg_data_, &current_call);

```

### CODE REVIEW

File: pjsip/include/pjsua-lib/pjsua.h [hunk 1]
```diff
@@ -1092,6 +1092,13 @@ typedef struct pjsua_call_setting
      */
     pjmedia_dir      media_dir[PJMEDIA_MAX_SDP_MEDIA];
 
+    /***
+     * Custom Call-ID to be sent out with INVITE. Overrides default
+     * Call-ID generated by dialog.
+     */
+
+    pj_str_t        user_call_id;
```
[the remark this answers is not in the document]
Comment (austinkottke) [reply pjsip/include/pjsua-lib/pjsua.h]: Changed to customCallId

File: pjsip/include/pjsua-lib/pjsua.h [hunk 2]
```diff
@@ -1092,6 +1092,13 @@ typedef struct pjsua_call_setting
      */
     pjmedia_dir      media_dir[PJMEDIA_MAX_SDP_MEDIA];
 
+    /***
+     * Custom Call-ID to be sent out with INVITE. Overrides default
+     * Call-ID generated by dialog.
```
[the remark this answers is not in the document]
Comment (austinkottke) [reply pjsip/include/pjsua-lib/pjsua.h:1097]: Added your notes to documentation

**Comment by austinkottke** [comment]
Commits have been added. Updated documentation to cpp api and c api to explicitly mention it is for outgoing calls only.

Changed the property to customCallId

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 3]
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
...
     }
 
+    /*
+     * Use pre defined Call-ID to be sent out with INVITE as opposed to using
+     * a randomly generated Call-ID
+     */
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%s)", dlg->call_id->id  ));
```
Comment (sauwming) [inline pjsip/src/pjsua-lib/pjsua_call.c:965]: `dlg->call_id->id` is of type `pj_str_t`. You need to use `%.*s` to print.

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 4]
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
...
     }
 
+    /*
+     * Use pre defined Call-ID to be sent out with INVITE as opposed to using
+     * a randomly generated Call-ID
+     */
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%.*s)", (int)dlg->call_id->id.slen, dlg->call_id->id  ));
```
Comment (sauwming) [inline pjsip/src/pjsua-lib/pjsua_call.c:965]: dlg->call_id->id.ptr

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 5]
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
         goto on_error;
     }
 
+    /*
+     * Use pre defined Call-ID to be sent out with INVITE as opposed to using
+     * a randomly generated Call-ID
+     */
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
```
Comment (sauwming) [inline pjsip/src/pjsua-lib/pjsua_call.c:964]: Need to duplicate the string using pj_strdup() as `opt` argument is const.

Comment (austinkottke) [reply pjsip/src/pjsua-lib/pjsua_call.c]: Ok I see - thank you for pointing that out.

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 6]
```diff
@@ -644,6 +644,7 @@ PJ_DEF(void) pjsua_call_setting_default(pjsua_call_setting *opt)
     pj_bzero(opt, sizeof(*opt));
     opt->flag = PJSUA_CALL_INCLUDE_DISABLED_MEDIA;
     opt->aud_cnt = 1;
+    opt->custom_call_id = pj_str("");
```
Comment (nanangizz) [inline pjsip/src/pjsua-lib/pjsua_call.c:647]: minor optimization: unnecessary? as the `bzero()` above should set `pj_str_t.len` to zero already.

File: pjsip/include/pjsua-lib/pjsua.h [hunk 7]
```diff
@@ -1092,6 +1092,15 @@ typedef struct pjsua_call_setting
      */
     pjmedia_dir      media_dir[PJMEDIA_MAX_SDP_MEDIA];
 
+    /**
+     * Custom Call-ID to be sent out with outgoing INVITE. Overrides default
+     * Call-ID generated by dialog.
+     *
+     * This setting will only be used when creating a new outgoing call via pjsua_call_make_call().
```
Comment (nanangizz) [inline pjsip/include/pjsua-lib/pjsua.h:1099]: - Better add info/warning that the library won't verify its validity (e.g: allowed chars, uniqueness).
- Column width should be < 80, for both, code & comment.
  Rather old school, but for consistency :)

File: pjsip/include/pjsua2/call.hpp [hunk 8]
```diff
@@ -333,6 +333,13 @@ struct CallSetting
      */
     MediaDirVector mediaDir;
 
+    /**
+     * User defined Call-ID to be sent out with outgoing INVITE
+     *
+     * This setting will only be used when creating a new outgoing call via Call::makeCall().
```
Comment (nanangizz) [inline pjsip/include/pjsua2/call.hpp:339]: Similar as above in `pjsua_call_setting`:
- Info/warning about no validity check.
- Col width > 80?

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 9]
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
...
     }
 
+    /*
+     * Use pre defined Call-ID to be sent out with INVITE as opposed to using
+     * a randomly generated Call-ID
+     */
+
+    if( opt->custom_call_id.slen > 0 ){
+        pj_strdup(dlg->pool, &dlg->call_id->id, &opt->custom_call_id);
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%.*s)", (int)dlg->call_id->id.slen, dlg->call_id->id.ptr  ));
```
Comment (nanangizz) [inline pjsip/src/pjsua-lib/pjsua_call.c:965]: col width > 80?

File: pjsip/include/pjsua-lib/pjsua.h (hunk 7 above)
Comment (austinkottke) [reply pjsip/include/pjsua-lib/pjsua.h]: ok sounds good :)

File: pjsip/src/pjsua-lib/pjsua_call.c (hunk 6 above)
Comment (austinkottke) [reply pjsip/src/pjsua-lib/pjsua_call.c]: Removed this. It appears unnecessary in my tests.

**Review by nanangizz — APPROVED** [review]

**Review by trengginas — APPROVED** [review]

**Review by sauwming — APPROVED** [review]

**Pull request merged**